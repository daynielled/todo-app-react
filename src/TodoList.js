import React, { useState, useEffect } from "react";
import Todo from "./Todo";
import NewTodoForm from "./NewTodoForm";

function TodoList() {
    const [todos, setTodos] = useState([]);

    // useEffect(() => {
    //     //Retrieve todos from localstorage on page load
    //     const savedTodos = JSON.parse(localStorage.getItem("todos")) || [];
    //     setTodos(savedTodos);
    // }, []);

    // useEffect(() => {
    //     //Save todos to localStorage whenever todos change
    //     localStorage.setItem("todos", JSON.stringify(todos));
    // }, [todos]);

    //Add a new todo
    const addNew = newTodo => {
        setTodos(todos => [...todos, newTodo]);
    };

     // update a todo with updatedTask
  const updateTodo = (id, updatedTask, completed) => {
    setTodos(todos =>
      todos.map(todo =>
        todo.id === id 
        ? {
            ...todo,
            task: updatedTask !== "" ? updatedTask : todo.task,
            completed: completed !== undefined ? completed : todo.completed,
          }
          : todo
      )
    );
  };

    //Remove a todo by id
    const removeTodo = (id) => {
        setTodos(todos => todos.filter(todo => todo.id !== id));
    };

    return (
        <div>
            <NewTodoForm addNewTodo={ addNew } />
            {todos.map((todo) => (
                <Todo
                    key={todo.id}
                    id={todo.id}
                    task={todo.task}
                    completed={todo.completed}
                    removeTodo={removeTodo}
                    updateTodo={updateTodo}
                />
            ))}
        </div>
    );
}

export default TodoList;