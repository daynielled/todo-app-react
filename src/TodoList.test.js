import React from "react";
import { render, fireEvent, screen, waitFor } from "@testing-library/react";
import TodoList from "./TodoList";

function addTodo(todoList, task = "write tests") {
  const taskInput = todoList.getByPlaceholderText("Enter Task");
  fireEvent.change(taskInput, { target: { value: task }});
  const submitButton = todoList.getByText("Add Task!");
  fireEvent.click(submitButton);
}

it("renders without crashing", function() {
  render(<TodoList />);
});

it("matches snapshot", function() {
  const { asFragment } = render(<TodoList />);
  expect(asFragment()).toMatchSnapshot();
});

it("can add a todo", function() {
  const list = render(<TodoList />);
  addTodo(list);

  // expect form to clear and todo to be on the page
  expect(list.getByPlaceholderText("Enter Task")).toHaveValue("");
  expect(list.getByText("write tests")).toBeInTheDocument();
  expect(list.getByText("Edit")).toBeInTheDocument();
  expect(list.getByText("Delete")).toBeInTheDocument();
  expect(list.getByText("Mark As Completed")).toBeInTheDocument();
});

  
