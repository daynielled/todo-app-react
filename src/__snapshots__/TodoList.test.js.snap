// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div>
      <form>
        <label
          for="task-1760b3d7-947b-40b0-bf2f-42bec03e0162"
        >
          Enter Task
        </label>
        <input
          id="task-1760b3d7-947b-40b0-bf2f-42bec03e0162"
          name="task"
          placeholder="Enter Task"
          type="text"
          value=""
        />
        <button>
          Add Task!
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
